### PR TITLE
[WIP] Filetype curapackage poc

### PR DESCRIPTION
**Pull request by ChrisTerBeke**
_(This PR has been opened for early comments and feedback)_

This PR adds support for .curapackage (extension can change!) files to libCharon. In order to do this, the following things have been done:

* Make an abstract OpenPackagingConvention filetype.
* Make UltimakerFormatPackage filetype a child of OpenPackagingConvention filetype.
* Make a new filetype called CuraPackage.
* Add unit tests for CuraPackage.
* Add the ability to define a `root_path` prefix to the `package.listPaths` method (e.g. to get all materials in `/resources/materials` within a CuraPackage).
* Quite a few bug fixes.

More things still in progress...

### CODE REVIEW

**Comment by ChrisTerBeke** [comment]
@LipuFei @ianpaschal @Appesteijn This is a start with defining .curapackage in libCharon that I was talking about. It'll help us document what a .curapackage actually looks like, which files are in there and how Cura/Cloud should read/write them. This will also help us generate the packages automatically based on the contents in GitHub.

**Comment by ChrisTerBeke** [comment]
The 2 unit tests that I've added now correctly create .curapackage files with contents in the expected places (one with 2 material files under `/files/resources/materials` and one with a `/Metadata/package.json` file that contains a `package_id` field with the given value).

This means that we can start defining the whole file structure and then fill it out in the `CuraPackage` class.

File: Charon/filetypes/OpenPackagingConvention.py [hunk 1]
```diff
@@ -406,7 +406,11 @@ def _readMetadata(self) -> None:
                 metadata_file = relationship.attrib["Target"]
                 if metadata_file not in self.zipfile.namelist():  # The metadata file is unknown to us.
                     continue
-                metadata = json.load(self.zipfile.open(metadata_file))
```
Comment (ChrisTerBeke) [inline Charon/filetypes/OpenPackagingConvention.py]: @awhiemstra How did this work before? `zipfile.open` doesn't even return something that `json.load` could interpret...

**Comment by ChrisTerBeke** [comment]
I'll fix all the typing stuff later (as well as the merge conflicts), more looking for general feedback on the idea of extending libCharon this way.

**Comment by nallath** [comment]
I'm in favour of changing it. We might consider also looking at libSavitar integration though as it's starting to move closer to what this lib is doing.

**Comment by ChrisTerBeke** [comment]
Ugh super annoying conflicts as I moved the main OPC implementation to an abstract filetype and there's many changes in the UFP filetype now.

File: Charon/filetypes/CuraPackage.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    # File aliases for quick and easy access.
+    aliases = OrderedDict([
+        (r"/materials", "/files/resources/materials")
+    ])
+
+    def getMaterials(self) -> List[str]:
+        """
+        Get all material files.
+        :return: A list of paths of materials.
+        """
```
Comment (sedwards2009) [inline Charon/filetypes/CuraPackage.py]: Doxygen style docs are the standard in Charon.

Comment (ChrisTerBeke) [reply Charon/filetypes/CuraPackage.py]: okay fine. It's just a bit annoying that PyCharm doesn't properly auto-complete those.... unless someone knows a setting/plugin for that?

**Comment by ChrisTerBeke** [comment]
Closing in favour of https://github.com/Ultimaker/libCharon/pull/6.

**Pull request closed without merging**